pragma-node/src/utils/ws.rs

```diff
@@ -237,75 +332,152 @@ where
-                        .map_err(|_| WebSocketError::ChannelClose)?;
-                    self.closed = true;
-                } else {
+                if self.exit.0.send(true).is_err() {
                     self.record_metric(Interaction::CloseConnection, Status::Error);
                 }
             }
+
             Message::Text(text) => {
-                let maybe_msg = serde_json::from_str::<T>(&text);
-                if let Ok(msg) = maybe_msg {
-                    self.last_activity = std::time::Instant::now();
-                    return Ok(Some(msg));
+                self.assert_client_message_size(text.len()).await?;
+
+                match serde_json::from_str::<T>(&text) {
+                    Ok(msg) => {
+                        self.last_activity = std::time::Instant::now();
+                        return Ok(Some(msg));
+                    }
+                    Err(e) => {
+                        self.send_err("Error parsing JSON into valid websocket request.")
+                            .await;
+                        return Err(WebSocketError::DecodingError(e.to_string()));
+                    }
                 }
-                tracing::error!("Failed to decode text message: {:?}", maybe_msg.err());
-                self.send_err(
-                    "⛔ Incorrect message. Please check the documentation for more information.",
-                )
-                .await;
-                return Err(WebSocketError::MessageDecode(text.to_string()));
             }
+
             Message::Binary(payload) => {
-                let maybe_msg = serde_json::from_slice::<T>(&payload);
-                if let Ok(msg) = maybe_msg {
-                    self.last_activity = std::time::Instant::now();
-                    return Ok(Some(msg));
+                self.assert_client_message_size(payload.len()).await?;
+
+                match serde_json::from_slice::<T>(&payload) {
+                    Ok(msg) => {
+                        self.last_activity = std::time::Instant::now();
+                        return Ok(Some(msg));
+                    }
+                    Err(e) => {
+                        self.send_err("Error parsing JSON into valid websocket request.")
+                            .await;
+                        return Err(WebSocketError::DecodingError(e.to_string()));
+                    }
                 }
-                self.send_err(
-                    "⛔ Incorrect message. Please check the documentation for more information.",
-                )
-                .await;
-                return Err(WebSocketError::MessageDecode(format!("{payload:?}")));
             }
+
             Message::Ping(_) => {
                 self.last_activity = std::time::Instant::now();
-                let _ = self.sender.send(Message::Pong(Default::default())).await;
+                let _ = self
+                    .send_sender
+                    .send(Message::Pong(Default::default()))
```

Comment:
we should send a pong with the same payload as the ping tho

https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#pings_and_pongs_the_heartbeat_of_websockets

Reply:
done